src/cli/subcommands/mpool_cmd.rs

```diff
@@ -1,16 +1,226 @@
 
+use std::str::FromStr;
+
+use crate::json::cid::vec::CidJsonVec;
+use crate::message::{Message, SignedMessage};
+use crate::rpc_client::{chain_ops::*, mpool_ops::*, state_ops::*, wallet_ops::*};
+use crate::shim::{address::Address, econ::TokenAmount};
+
+use ahash::{HashMap, HashSet};
 use clap::Subcommand;
+use fvm_ipld_encoding::Cbor;
+use num::BigInt;
 
-use super::Config;
+use super::{handle_rpc_err, Config};
 
 #[derive(Debug, Subcommand)]
-pub enum MpoolCommands {}
+pub enum MpoolCommands {
+    /// Get pending messages
+    Pending {
+        /// Print pending messages for addresses in local wallet only
+        #[arg(long)]
+        local: bool,
+        /// Only print `CIDs` of messages in output
+        #[arg(long)]
+        cids: bool,
+        /// Return messages to a given address
+        #[arg(long)]
+        to: Option<String>,
+        /// Return messages from a given address
+        #[arg(long)]
+        from: Option<String>,
+    },
+    /// Print mempool stats
+    Stat {
+        /// Number of blocks to look back for minimum `basefee`
+        #[arg(long, default_value = "60")]
+        basefee_lookback: u32,
+        /// Print stats for addresses in local wallet only
+        #[arg(long)]
+        local: bool,
+    },
+}
+
+fn to_addr(value: &Option<String>) -> anyhow::Result<Option<Address>> {
+    Ok(value.as_ref().map(|s| Address::from_str(s)).transpose()?)
+}
 
 impl MpoolCommands {
-    pub fn run(&self, _config: Config) -> anyhow::Result<()> {
-        // match self {}
-        Ok(())
+    pub async fn run(&self, config: Config) -> anyhow::Result<()> {
+        match self {
+            Self::Pending {
+                local,
+                cids,
+                to,
+                from,
+            } => {
+                let to = to_addr(to)?;
+                let from = to_addr(from)?;
+
+                let messages = mpool_pending((CidJsonVec(vec![]),), &config.client.rpc_token)
+                    .await
+                    .map_err(handle_rpc_err)?;
+
+                let local_addrs = if *local {
+                    let response = wallet_list((), &config.client.rpc_token)
+                        .await
+                        .map_err(handle_rpc_err)?;
+                    Some(HashSet::from_iter(response.iter().map(|addr| addr.0)))
+                } else {
+                    None
+                };
+
+                let filtered_messages = messages.iter().filter(|msg| {
+                    local_addrs
+                        .as_ref()
+                        .map(|addrs| addrs.contains(&msg.0.from()))
+                        .unwrap_or(true)
+                        && to.map(|addr| msg.0.to() == addr).unwrap_or(true)
+                        && from.map(|addr| msg.0.from() == addr).unwrap_or(true)
+                });
+                for msg in filtered_messages {
+                    if *cids {
+                        println!("{}", msg.0.cid().unwrap());
```

Comment:
is it safe?

Reply:
Yes, by construction `mpool_pending` will ensure that message cids are ok.